x-pack/plugins/lens/public/state_management/lens_slice.ts

```diff
@@ -282,6 +301,100 @@ export const makeLensReducer = (storeDeps: LensStoreDeps) => {
           ? activeVisualization.clearLayer(state.visualization.state, layerId)
           : activeVisualization.removeLayer(state.visualization.state, layerId);
     },
+    [changeIndexPattern.type]: (
+      state,
+      {
+        payload,
+      }: {
+        payload: {
+          visualizationIds?: string;
+          datasourceIds?: string;
+          layerId?: string;
+          indexPatternId: string;
+          dataViews: Pick<DataViewsState, 'indexPatterns'>;
+        };
+      }
+    ) => {
+      const { visualizationIds, datasourceIds, layerId, indexPatternId, dataViews } = payload;
+      const newState: Partial<LensAppState> = { dataViews: { ...state.dataViews, ...dataViews } };
```

Comment:
I think I realized this problem - on changing the index pattern, we don't update the list of index pattern refs, but if it has been created on the spot this is necessary. At this point we could make sure that the new index pattern is listed in the refs and add it if that's not the case.

Reply:
Right. I've added a loading check in the top nav change index pattern function to reload all refs in case the target id is missing in the refs cache.